src/burn/engine/apply.cpp

```diff
@@ -745,6 +771,36 @@ extern "C" HRESULT ApplyExecute(
             continue;
         }
 
+		// Transaction end/start
+		if (BURN_EXECUTE_ACTION_TYPE_ROLLBACK_BOUNDARY == pExecuteAction->type)
+		{
+			// End previous transaction
+			if (fInTransaction)
+			{
+				LogString(REPORT_STANDARD, "Committing MSI transaction\n");
+				hr = DoMsiCommitTransaction(&context, pEngineState);
+				ExitOnFailure(hr, "Failed committing an MSI transaction");
+				fInTransaction = FALSE;
+			}
+
+			// Start New transaction
+			if (!fInTransaction && pExecuteAction->rollbackBoundary.pRollbackBoundary && pExecuteAction->rollbackBoundary.pRollbackBoundary->fTransaction)
+			{
+				// Transactions don't go together with DisableRollback.
+				if (pEngineState->fDisableRollback)
+				{
+					LogString(REPORT_STANDARD, "Ignoring Transaction flag due to DisableRollback flag\n");
+				}
+				else
+				{
+					LogString(REPORT_STANDARD, "Starting a new MSI transaction\n");
+					hr = DoMsiBeginTransaction(&context, pEngineState);
```

Comment:
If we're really going to bail if the transaction can't be started and not fallback to burn's traditional rollback logic, then we should add new plan actions for starting/completing/rolling back the transaction instead of doing it in the BURN_EXECUTE_ACTION_TYPE_ROLLBACK_BOUNDARY action, and not even schedule rollback actions for each individual package.